devito/petsc/iet/passes.py

```diff
@@ -0,0 +1,366 @@
+        solver_objs['snes'], 'SNESKSPONLY'])])
+
+    global_x = Call(petsc_call, [Call('DMCreateGlobalVector', arguments=[
+        objs['da'], Byref(solver_objs['x_global'])])])
+
+    local_x = Call(petsc_call, [Call('DMCreateLocalVector', arguments=[
+        objs['da'], Byref(solver_objs['x_local'])])])
+
+    global_b = Call(petsc_call, [Call('DMCreateGlobalVector', arguments=[
+        objs['da'], Byref(solver_objs['b_global'])])])
+
+    local_b = Call(petsc_call, [Call('DMCreateLocalVector', arguments=[
+        objs['da'], Byref(solver_objs['b_local'])])])
+
+    snes_get_ksp = Call(petsc_call, [Call('SNESGetKSP', arguments=[
+        solver_objs['snes'], Byref(solver_objs['ksp'])])])
+
+    vec_replace_array = Call(petsc_call, [
+        Call('VecReplaceArray', arguments=[solver_objs[
+            'x_local'], FieldFromPointer(target._C_field_data, target._C_symbol)])])
+
+    ksp_set_tols = Call(petsc_call, [Call('KSPSetTolerances', arguments=[
+        solver_objs['ksp'], solver_params['ksp_rtol'], solver_params['ksp_atol'],
+        solver_params['ksp_divtol'], solver_params['ksp_max_it']])])
+
+    ksp_set_type = Call(petsc_call, [Call('KSPSetType', arguments=[
+        solver_objs['ksp'], linear_solver_mapper[solver_params['ksp_type']]])])
+
+    ksp_get_pc = Call(petsc_call, [Call('KSPGetPC', arguments=[
+        solver_objs['ksp'], Byref(solver_objs['pc'])])])
+
+    pc_set_type = Call(petsc_call, [Call('PCSetType', arguments=[
+        solver_objs['pc'], linear_solver_mapper[solver_params['pc_type']]])])
+
+    ksp_set_from_ops = Call(petsc_call, [Call('KSPSetFromOptions', arguments=[
+        solver_objs['ksp']])])
+
+    return tuple([snes_create, snes_set_dm, create_matrix, snes_set_type,
+                  global_x, local_x, global_b, local_b, snes_get_ksp,
+                  vec_replace_array, ksp_set_tols, ksp_set_type, ksp_get_pc,
+                  pc_set_type, ksp_set_from_ops])
+
+
+def create_matvec_callback(target, body, solver_objs, objs, struct):
+
+    # There will be 2 PETScArrays within the body
+    petsc_arrays = [i.function for i in FindSymbols('indexedbases').visit(body)
+                    if isinstance(i.function, PETScArray)]
+
+    # There will only be one PETScArray that is written to within this body and
+    # one PETScArray which corresponds to the 'seed' vector
+    petsc_arr_write, = FindSymbols('writes').visit(body)
+    petsc_arr_seed, = [i for i in petsc_arrays if i.function != petsc_arr_write.function]
+
+    # Struct needs to be defined explicitly here since CompositeObjects
+    # do not have 'liveness'
+    defn_struct = Definition(struct)
+
+    mat_get_dm = Call(petsc_call, [Call('MatGetDM', arguments=[
+        solver_objs['Jac'], Byref(objs['da'])])])
+
+    dm_get_app_context = Call(petsc_call, [Call('DMGetApplicationContext', arguments=[
+        objs['da'], Byref(struct._C_symbol)])])
+
+    dm_get_local_xvec = Call(petsc_call, [Call('DMGetLocalVector', arguments=[
```

Comment:
Some repetition in this section. Would it be worth creating some utility to build these to improve readability?

Reply:
I agree to improve readability you may add simple utility functions creating `Call(... [Call(....)])`